lsm303agr/lsm303agr.go

```diff
@@ -0,0 +1,196 @@
+func (d *Device) Configure(cfg Configuration) {
+
+	if cfg.AccelDataRate != 0 {
+		d.AccelDataRate = cfg.AccelDataRate
+	} else {
+		d.AccelDataRate = ACCEL_DATARATE_100HZ
+	}
+
+	if cfg.AccelPowerMode != 0 {
+		d.AccelPowerMode = cfg.AccelPowerMode
+	} else {
+		d.AccelPowerMode = ACCEL_POWER_NORMAL
+	}
+
+	if cfg.AccelRange != 0 {
+		d.AccelRange = cfg.AccelRange
+	} else {
+		d.AccelRange = ACCEL_RANGE_2G
+	}
+
+	if cfg.MagPowerMode != 0 {
+		d.MagPowerMode = cfg.MagPowerMode
+	} else {
+		d.MagPowerMode = MAG_POWER_NORMAL
+	}
+
+	if cfg.MagDataRate != 0 {
+		d.MagDataRate = cfg.MagDataRate
+	} else {
+		d.MagDataRate = MAG_DATARATE_10HZ
+	}
+
+	if cfg.MagSystemMode != 0 {
+		d.MagSystemMode = cfg.MagSystemMode
+	} else {
+		d.MagSystemMode = MAG_SYSTEM_CONTINUOUS
+	}
+
+	cmd := []byte{0}
+
+	cmd[0] = byte(d.AccelDataRate<<4 | d.AccelPowerMode | 0x07)
+	d.bus.WriteRegister(uint8(d.AccelAddress), ACCEL_CTRL_REG1_A, cmd)
+
+	cmd[0] = byte(0x80 | d.AccelRange<<4)
+	d.bus.WriteRegister(uint8(d.AccelAddress), ACCEL_CTRL_REG4_A, cmd)
+
+	cmd[0] = byte(0xC0)
+	d.bus.WriteRegister(uint8(d.AccelAddress), TEMP_CFG_REG_A, cmd)
+
+	cmd[0] = byte(0x80 | d.MagPowerMode<<4 | d.MagDataRate<<2 | d.MagSystemMode)
+	d.bus.WriteRegister(uint8(d.MagAddress), MAG_MR_REG_M, cmd)
+
+}
+
+/* read raw acceleration data from all axis, 1000+ = 1g */
+func (d *Device) ReadAcceleration() (x int16, y int16, z int16) {
+
+	data1, data2, data3, data4, data5, data6 := []byte{0}, []byte{0}, []byte{0}, []byte{0}, []byte{0}, []byte{0}
+	d.bus.ReadRegister(uint8(d.AccelAddress), ACCEL_OUT_X_H_A, data1)
+	d.bus.ReadRegister(uint8(d.AccelAddress), ACCEL_OUT_X_L_A, data2)
+	d.bus.ReadRegister(uint8(d.AccelAddress), ACCEL_OUT_Y_H_A, data3)
+	d.bus.ReadRegister(uint8(d.AccelAddress), ACCEL_OUT_Y_L_A, data4)
+	d.bus.ReadRegister(uint8(d.AccelAddress), ACCEL_OUT_Z_H_A, data5)
+	d.bus.ReadRegister(uint8(d.AccelAddress), ACCEL_OUT_Z_L_A, data6)
+
+	range_factor := int16(0)
+	switch d.AccelRange {
+	case ACCEL_RANGE_2G:
+		range_factor = 1
+	case ACCEL_RANGE_4G:
+		range_factor = 2
+	case ACCEL_RANGE_8G:
+		range_factor = 4
+	case ACCEL_RANGE_16G:
+		range_factor = 12 // the readings in 16G are a bit off
+	}
+
+	x = int16((uint16(data1[0])<<8 | uint16(data2[0]))) >> 4 * range_factor
+	y = int16((uint16(data3[0])<<8 | uint16(data4[0]))) >> 4 * range_factor
+	z = int16((uint16(data5[0])<<8 | uint16(data6[0]))) >> 4 * range_factor
+	return
+}
+
+/* read pitch/roll degrees */
+func (d *Device) ReadPitchRoll() (pitch float32, roll float32) {
+
+	x, y, z := d.ReadAcceleration()
+	xf, yf, zf := float64(x), float64(y), float64(z)
+
+	pitch = float32(math.Round(math.Atan2(yf, math.Sqrt(math.Pow(xf, 2)+math.Pow(zf, 2)))*(180/math.Pi)*100) / 100)
+	roll = float32(math.Round(math.Atan2(xf, math.Sqrt(math.Pow(yf, 2)+math.Pow(zf, 2)))*(180/math.Pi)*100) / 100)
+	return
+
+}
+
+/* read magnetic field level from all axis */
+func (d *Device) ReadMagneticField() (x int16, y int16, z int16) {
```

Comment:
In similar fashion to my previous comment, this function would be `func (d *Device) ReadRotation() (x int32, y int32, z int32)` and return values in µ°/s (micro-degrees/sec).

Reply:
I think it returns microgauss. In MakeCode it's microtesla.

Reply:
One purpose of these APIs is to give a consistent interface for different drivers. It is not a consistent interface if every driver has a different scale (even if the signature is the same). There are many benefits to a consistent interface, such as that it is easier to learn. Therefore, I would suggest matching the existing drivers for consistency.

Reply:
Oh, I misread the previous post. I thought it was about ReadMagneticField(). It's ReadRotation().

Will try to fix it.

Reply:
Note: I can convert it to microdegree since the magnetic sensor is not a gyroscope. It sense the "direction" by Earth's magnetic field, not the rotating motion.

Or should I name it ReadCompass()?